### DISCUSSION TITLE
Zig API 设计理念与抽象

### DISCUSSION THREAD

**Discussion by jiacai2050** [zigcc/forum#68] [General]
这次讨论的话题来自 [What is the Zig philosophy on APIs and abstraction?](https://www.reddit.com/r/Zig/comments/17xd46v/what_is_the_zig_philosophy_on_apis_and_abstraction/) 这个 Reddit 帖子，讨论 Zig 的在 API 设计上的哲学、理念。一个有意思的点是 Zig 不支持私有的字段，全部都是 public 的。Andrew 在 [Proposal: Private Fields #9909](https://github.com/ziglang/zig/issues/9909#issuecomment-942686366) 这个 issue 里面讨论过原因，主要根据：

-   一个结构体的抽象，很难保证不泄漏，比如一个类型的 align、size，一个函数是否可以在 comptime 执行
-   一个包的兼容性，应该由文档来解释
-   增加私有字段，会增加语言的复杂度，而且这种复杂性本身是完全可以避免的

**Comment by jinzhongjia** [comment]
目前的字段完全公共很依赖于文档本身的说明，有趣的是现在zig官方文档中仍存在着不少 commit 根本不写注释说明或者非常简短，让人看得云里雾里

**Reply by jiacai2050** [reply]
没有明确说明，就认为是 public 就好了，一般不想暴露出来的变量名会单独注明。

**Comment by jinzhongjia** [comment]
其实我认为或许可以在注释层面额外增加一层标记是否为私有，这样子可以让zls本身给出信息提示，但在语言实现时又没有额外的复杂性增加

**Comment by zhuyadong** [comment]
个人认为，自己定个规则 ，比如私有的变量前置下划线就行了。以我在c#中的经验来看，c#有private关键字，但是大家写代码的时候，还是会把private的成员命名成'_xxx'这样，既然如此，何必多个private声明出来 ，变量名自包含不是更简单。
